Columnar: support relation options with ALTER TABLE.

jeff-davis opened the pull request:
Use ALTER TABLE ... SET/RESET to specify relation options rather than
alter_columnar_table_set() and alter_columnar_table_reset().
    
Not only is this more ergonomic, but it also allows better integration
because it can be treated like DDL on a regular table. For instance,
citus can use its own ProcessUtility_hook to distribute the new
settings to the shards.

DESCRIPTION: Columnar: support relation options with ALTER TABLE.

thanodnl commented on src/backend/columnar/columnar_metadata.c:
```diff
@@ -218,6 +219,117 @@ InitColumnarOptions(Oid regclass)
+void
+SetColumnarRelOptions(RangeVar *rv, List *reloptions)
+{
+	ColumnarOptions options;
```
we tend to zero initialize all variables

```suggestion
	ColumnarOptions options = { 0 };
```

jeff-davis replied:
Done.

thanodnl commented on src/backend/columnar/columnar_metadata.c:
```diff
@@ -218,6 +219,117 @@ InitColumnarOptions(Oid regclass)
+{
+	List *columnarOptions = NIL;
+	List *otherOptions = NIL;
+	ListCell *lc;
```
same as below

```suggestion
	ListCell *lc = NULL;
```

jeff-davis replied:
Done.

thanodnl commented on src/backend/columnar/columnar_metadata.c:
```diff
@@ -218,6 +219,117 @@ InitColumnarOptions(Oid regclass)
+		{
+			options.compressionLevel = (elem->arg == NULL) ?
+				columnar_compression_level : defGetInt64(elem);
+		}
```
I was expecting a warning or error here if an unkown columnar option would be tried to set. Do we sanitize the option list elsewhere? Or is there an other good reason to allow arbitrary `columnar.options` to be set?

thanodnl replied:
As expected the current behaviour doesn't error/warn when we set a non-existing columnar option. Instead it is silently ignored:

```
[local] citus@citus:9700-18928=# ALTER TABLE c SET (columnar.barbaz = fooz);
ALTER TABLE
Time: 76.822 ms
```

Would this be desired behaviour?

default behaviour of heap is:
```
[local] citus@citus:9700-18928=# ALTER TABLE t1 SET (something_bogus = 42);
ERROR:  unrecognized parameter "something_bogus"
Time: 0.540 ms
```

I guess we would want to retain such behaviour.

jeff-davis replied:
Fixed.

thanodnl commented on src/backend/columnar/columnar_tableam.c:
```diff
@@ -2126,9 +2128,87 @@ ColumnarProcessUtility(PlannedStmt *pstmt,
 	PrevProcessUtilityHook_compat(pstmt, queryString, false, context,
 								  params, queryEnv, dest, completionTag);
+
+	if (columnarRangeVar != NULL && columnarOptions != NIL)
```
list should generally not be compared to `NIL` iirc, instead `list_length() > 0`. Unless there is a semantic difference between no list, and a list of length 0.

jeff-davis replied:
Hmm. That's all over the postgres code base, is this a citus coding standard thing?

thanodnl replied:
looking at the postgres source I find the following comment:
```
/*
 * The *only* valid representation of an empty list is NIL; in other
 * words, a non-NIL list is guaranteed to have length >= 1.
 */
#define NIL						((List *) NULL)
```

So I guess this is ok, probably messed up the notion of empty lists in golang.

thanodnl commented on src/backend/columnar/columnar_tableam.c:
```diff
@@ -2126,9 +2128,87 @@ ColumnarProcessUtility(PlannedStmt *pstmt,
 
 		RelationClose(rel);
 	}
+	else if (IsA(parsetree, CreateStmt))
```
I think now would be a great moment to change the if/elseif pairs here to a switch statement.

jeff-davis replied:
Done.

thanodnl commented on src/backend/columnar/columnar_tableam.c:
```diff
@@ -2126,9 +2128,87 @@ ColumnarProcessUtility(PlannedStmt *pstmt,
+				if (isColumnar || toColumnar)
+				{
+					columnarRangeVar = alterTableStmt->relation;
+					alterTableCmd->def = (Node *) ExtractColumnarRelOptions(options,
+																			&columnarOptions);
```
We are looping over `AlterTableCmd`'s. With the current implementation of `ExtractColumnarRelOptions` we overwrite the `columnarOptions` variable instead of appending the options to the list.

Could there be a posibility of hitting the `ExtractColumnarRelOptions` call here more then once? In that case we will _only_ have the last options extracted.

Alternatively we could change the behaviour of `ExtractColumnarRelOptions` to _not_ swap the lists, but instead `lappend` directly to the out parameter.

If this could only ever be hit once, would it make sense to break out of the loop? Guess that is not a possibility due to the PG15 block below. Then an `elog` when we hit this code wile `columnarOptions` is not empty anymore would help guard against undesired behaviour.

jeff-davis replied:
Fixed (and tested). New settings are appended to the list, and the last update wins.

thanodnl commented on src/backend/columnar/columnar_metadata.c:
```diff
@@ -218,6 +219,117 @@ InitColumnarOptions(Oid regclass)
+				char *compressionStr = defGetString(elem);
+				CompressionType compressionType = ParseCompressionType(compressionStr);
+
+				if (options.compressionType == COMPRESSION_TYPE_INVALID)
```
when setting the compression type to an unknown/invalid compression type, currently the backend crashes. eg.
```sql
ALTER TABLE c SET (columnar.compression = fooz);
```
Causes a crash when `c` is using the columnar access method.

I think it is due to the comparison we are doing here.
We will either need to set `options.compressionType` directly above, or we will want to compare against `compressionType` here.

This also indicates we don't have a test for setting the option to something bogus. Might be a good moment to add one.

jeff-davis replied:
Fixed. We do have one test for that, which caught the problem. Thank you.

thanodnl reviewed (changes requested):
Exciting to see this change! Feels even more native.

Would it make sense to deprecate the UDF we had before? Especially with Citus 11 coming up?

@marcocitus, how would we go in deprecating the udf? Could we simply remove it, given the major citus version bump comming up? Or do you want to keep it till our next major? Possibly with a warning message now?

@jeff-davis Most important change would prevent the crash when setting the compression method to something bogus.

But I also feel strongly about error-ing on unknown options, inline with heap.

jeff-davis commented:
> Would it make sense to deprecate the UDF we had before? Especially with Citus 11 coming up?
> 
> @marcocitus, how would we go in deprecating the udf? Could we simply remove it, given the major citus version bump comming up? Or do you want to keep it till our next major? Possibly with a warning message now?

I'd probably leave the function versions in for a bit, but maybe we could make them plpgsql instead of C functions?

> @jeff-davis Most important change would prevent the crash when setting the compression method to something bogus.

Fixed.

> But I also feel strongly about error-ing on unknown options, inline with heap.

Done.

jeff-davis commented:
I rewrote `alter_columnar_table_(re)set()` in `plpgsql` that just calls `ALTER TABLE ... (RE)SET`, and removed the dead code related to the columnar options hook. I think this PR is ready.

marcocitus commented on src/backend/columnar/sql/udfs/alter_columnar_table_reset/10.0-1.sql:
```diff
@@ -5,8 +5,39 @@ CREATE OR REPLACE FUNCTION pg_catalog.alter_columnar_table_reset(
     RETURNS void
-    LANGUAGE C
-AS 'MODULE_PATHNAME', 'alter_columnar_table_reset';
+    LANGUAGE plpgsql AS
```
As a policy, we generally do not change past SQL files, since there's a high risk of diverging schemas between upgrades vs. new installs. Instead we always write a new version (11.1-1.sql) and replay all scripts for new installs. In many cases this also makes it easier to write a correct downgrade script, because you can simply #include the previous version from the downgrade script.

Eventually, once we have phased out support for an older PostgreSQL version, we can collapse all SQL scripts up to the lowest version that supports one of the currently supported PostgreSQL versions.

For instance, Citus 9.5 is the first version that supported PostgreSQL 13, and PostgreSQL 13 is the lowest version supported by Citus 11.0. Hence, we can collapse all scripts from 8.0-1 up to 9.5-1 into a single script, because there is no scenario in which a user could upgrade directly from 9.4 or older to 11.0. When we collapse, we can also remove redundant creations.

jeff-davis replied:
Done, thank you.

thanodnl commented on src/backend/columnar/columnar_tableam.c:
```diff
@@ -2325,219 +2470,38 @@ ColumnarCheckLogicalReplication(Relation rel)
-	table_close(rel, NoLock);
-
-	PG_RETURN_VOID();
+	ereport(ERROR, (errmsg("alter_columnar_table_set is deprecated"),
+					errhint("Use ALTER TABLE ... SET instead.")));
```
nit; given this code should not be accessible anymore, would an `elog` make more sense?

jeff-davis replied:
Fixed.

thanodnl commented on src/backend/columnar/columnar_tableam.c:
```diff
@@ -2104,31 +2164,116 @@ ColumnarProcessUtility(PlannedStmt *pstmt,
 		}
 
-		RelationClose(rel);
+		case T_CreateTableAsStmt:
+		{
+			CreateTableAsStmt *createTableAsStmt = castNode(CreateTableAsStmt, parsetree);
+			IntoClause *into = createTableAsStmt->into;
+			bool no_op = false;
+
+			if (createTableAsStmt->if_not_exists)
+			{
+				Oid existing_relid;
+
+				/* use same check as transformCreateStmt */
+				(void) RangeVarGetAndCheckCreationNamespace(
+					into->rel, AccessShareLock, &existing_relid);
+
+				no_op = OidIsValid(existing_relid);
+			}
+
+			if (!no_op && into->accessMethod != NULL &&
+				!strcmp(into->accessMethod, COLUMNAR_AM_NAME))
+			{
+				columnarRangeVar = into->rel;
+				into->options = ExtractColumnarRelOptions(into->options,
+														  &columnarOptions);
+			}
+			break;
+		}
```
When creating a columnar table from `CREATE TABLE ... AS ...` I notice that the query still gets executed, even if one of the columnar options cause an error. Not a huge problem, I just wonder if it would make sense to throw errors before dispatching to the original utility hook.

```
[local] citus@citus:9700-11392=# CREATE TABLE c3 USING columnar WITH (columnar.chunk_group_row_limit = 1) AS SELECT * FROM t1;
NOTICE:  issuing SELECT a, b, c, d FROM citus.t1_102016 t1 WHERE true
DETAIL:  on server citus@localhost:9701 connectionId: 5
NOTICE:  issuing SELECT a, b, c, d FROM citus.t1_102017 t1 WHERE true
DETAIL:  on server citus@localhost:9702 connectionId: 6
NOTICE:  issuing SELECT a, b, c, d FROM citus.t1_102018 t1 WHERE true
DETAIL:  on server citus@localhost:9701 connectionId: 5
NOTICE:  issuing SELECT a, b, c, d FROM citus.t1_102019 t1 WHERE true
DETAIL:  on server citus@localhost:9702 connectionId: 6
ERROR:  chunk group row count limit out of range
HINT:  chunk group row count limit must be between 1000 and 100000
Time: 44.518 ms
```

I guess this is due to the error being thrown from `SetColumnarRelOptions`, and I think most of the time it is ok to throw the error late, as the work done is not much.

However, `CREATE TABLE ... AS ...` and, in pg15 changing the table access methods could be somewhat heavy IO operations.

Would it make sense to have a precheck of the columnar options before we dispatch `PrevProcessUtilityHook_compat`?

jeff-davis replied:
Good catch!

Fixed by refactoring. Worth it because it seems like a normal use case. Though with #5927, we have another related issue to fix.

thanodnl commented on src/backend/distributed/shared_library_init.c:
```diff
@@ -388,19 +385,15 @@ _PG_init(void)
-	Assert(*ColumnarTableSetOptions_hook_ptr != NULL);
-
-	**ColumnarTableSetOptions_hook_ptr = ColumnarTableSetOptionsHook;
+	PrevProcessUtility = (ProcessUtility_hook != NULL) ?
+						 ProcessUtility_hook : standard_ProcessUtility;
```
For Citus we introduce `PrevProcessUtility`, while for columnar we have `PrevProcessUtilityHook`. Both names are subtly different. I wonder if it would make sense to have specific `Columnar` and `Citus` in their names to avoid confusion.

I understand that both extensions are linked differently, and even the names are different. However it got me confused slightly while diving deep in the code that a) they were different, and b) which one is for Citus and Columnar. Mostly the file they are declared in dictates the difference, which was a bit hard, given they live in the same repo

jeff-davis replied:
Adding "citus" or "columnar" to the name makes it more confusing to me, because it either implies that we know which extension the previous hook belongs to (which is a layering violation), or that it belongs to the current extension (which it doesn't).

Looking in `contrib/`, it appears that `prev_ProcessUtility` or `next_ProcessUtility` are the preferred names. I guess the "prev" and "next" are supposed to imply when we are calling it relative to our own code... though I had always read it as "preexisting_ProcessUtility". Honestly not quite sure what to do here and I don't want to over-think it right now.

thanodnl reviewed (approved):
Looks good for merging.

Thanks for the changes on the sql definitions, good catch of @marcocitus.

Left a couple of comments that you could take a look at. But am good with merging as is.
